autopush/router/webpush.py

```diff
@@ -109,32 +108,6 @@ def route_notification(self, notification, uaid_data):
                                   response_body="Retry Request",
                                   errno=201)
 
-        # - Lookup client
-        #   - Success (node found): Notify node of new notification
-        #     - Success: Done, return 200
-        #     - Error (no client): Done, return 202
-        #     - Error (no node): Clear node entry
-        #       - Both: Done, return 202
-        #   - Success (no node): Done, return 202
-        #   - Error (db error): Done, return 202
-        #   - Error (no client) : Done, return 404
-        try:
-            # is this call redundant? We already get uaid_data passed in
-            uaid_data = yield deferToThread(router.get_uaid, uaid)
-        except JSONResponseError:
-            returnValue(self.stored_response(notification))
-        except ItemNotFound:
-            self.metrics.increment("updates.client.deleted")
-            raise RouterException("User was deleted",
-                                  status_code=410,
-                                  response_body="Invalid UAID",
-                                  log_exception=False,
-                                  errno=105)
-
-        # Verify there's a node_id in here, if not we're done
-        node_id = uaid_data.get("node_id")
-        if not node_id:
```

Comment:
Couldn't there still be a case of a node_id of None here?

Reply:
Also we may have successfully called clear_node above and fell through here with a now stale node_id

Reply:
Hrm, I think this might be a different bug, thanks!

We fetch the `uaid_data` in [the marshmallow code](https://github.com/mozilla-services/autopush/blob/master/autopush/web/webpush.py#L85), which should contain the node_id. We don't check if that node_id is valid. We then try to [send to that node_id](https://github.com/mozilla-services/autopush/blob/master/autopush/router/webpush.py#L84), and if that fails with a `connection` error, we [return a `delivered_response`](https://github.com/mozilla-services/autopush/blob/master/autopush/router/webpush.py#L155), which I don't believe actually saves the message. Does the code seem to suggest that to you too? 

I believe that we need to check for node_id earlier, and only attempt delivery if it is present, as well as make sure that we save messages that either do not have a node_id assigned, or have failed to be delivered. Correct?